Add support for parsing auxiliary images

johncf opened the pull request:
Fixes (partially) #117 .

Changes proposed in this pull request:

 * Added a property `aux_image_ids` to `CtxImage` that returns a list of non-depth non-alpha auxiliary image IDs.
 * Added a method `get_aux_type(aux_id)` to `CtxImage` that returns a string representing the `aux_type` of the given `aux_id`.
 * Added a key `aux` to `HeifImage.info` dict which is also a dict mapping from `aux_type` strings to a list of `aux_id`s.
 * Added a method `get_aux_image(aux_id)` to `CtxImage` that returns a `CtxImage` corresponding to the given `aux_id`.
 * Added a method `get_aux_image(aux_id)` to `HeifImage` which returns an instance of a new class `HeifAuxImage` wrapping the above `CtxImage`, after ensuring that the auxiliary image is 8-bit monochrome.
 * Added a method `get_aux_image(aux_id)` to `HeifFile` which calls the same method of the primary image.

bigcat88 commented on pillow_heif/heif.py:
```diff
@@ -127,17 +127,19 @@ def __init__(self, c_image):
-        _bytes = f"{len(self.data)} bytes" if self._data or isinstance(self._c_image, MimCImage) else "no"
         return f"<{self.__class__.__name__} {self.size[0]}x{self.size[1]} {self.mode}>"
 
-    def to_pillow(self) -> Image.Image:
```
what was the reason to remove this?

bigcat88 commented:
After this line https://github.com/bigcat88/pillow_heif/blob/b4b5d642acb224ba2b3569e07e38cd3bdee365f7/pillow_heif/misc.py#L480 
should be added `self.aux_image_ids: list = []` 
at least for tests to start and not fail on the first one

bigcat88 commented:
second:

GitHub was right, please return `to_pillow` implementation to the `HeifDepthImage` or **test_depth_image** will fail.

that step is need to copy `info` dict. (at least for now)

as I understand we do not need any additional metadata for AUX the images so in `HeifAuxImage` we does not need to overload `to_pillow` method, but for `HeifDepthImage` need to keep old behaviour to not break anything.

---

third:

please add some tests for aux images. It will be nice to have coverage for all code except `raise NotImplementedError` 


---

in general it looks fine to me such implementation.

johncf commented:
Thanks for the review. I'll make those changes later today or tomorrow. I wanted to get your opinion on a slightly different way of representing this auxiliary info. From a usage perspective, people will likely have an aux-type in mind to access. So with what I proposed above, they'll have to iterate through all the aux IDs and check for that aux-type. How about having `heif_image.info["aux"]` map to a dict from `aux_type` to a list of `aux_id`s?

So I can do:
```python
aux_type = "urn:com:apple:photo:2020:aux:hdrgainmap"
assert aux_type in heif_image.info["aux"]
aux_id = heif_image.info["aux"][aux_type][0]
hdrgainmap = heif_image.get_aux_image(aux_id)
```
instead of something like:
```python
aux_type = "urn:com:apple:photo:2020:aux:hdrgainmap"
aux_ids = [aux_id for aux_id, metadata in heif_image.info["aux"].items()
                  if metadata["type"] == aux_type]
assert len(aux_ids) > 0
hdrgainmap = heif_image.get_aux_image(aux_ids[0])
```

Side note: Please take a look at #298 too

bigcat88 commented:
> What do you think?

Personally, I am satisfied with both methods, but I am more interested in how I can get these aux images when using Pillow?

johncf commented:
> GitHub was right, please return `to_pillow` implementation to the `HeifDepthImage` or **test_depth_image** will fail.
> 
> that step is need to copy `info` dict. (at least for now)

You're right! That was a mistake on my end. I had misunderstood the code when I made that change.

> as I understand we do not need any additional metadata for AUX the images so in `HeifAuxImage` we does not need to overload `to_pillow` method, but for `HeifDepthImage` need to keep old behaviour to not break anything.

True, but I've added an overloaded method for `HeifAuxImage` too since it might be useful to check `info["aux_type"]` later on.

> please add some tests for aux images.

I'll add tests soon, and I have a sample heic image that I took on my iphone (having hdrgainmap) and cropped so the size is ~230KB. Shall I simply add it to the repo at `tests/images/heif_other`? (Does this repo use Git LFS?)

> but I am more interested in how I can get these aux images when using Pillow?

Right now, only registering a pillow load hook will not be enough to access AUX images since they're not available directly from `info`. One will need to use `pillow_heif` directly something like this:

```
heif_file = pillow_heif.open_heif("hdr-sample.heic")
aux_id = heif_file.info["aux"]["urn:com:apple:photo:2020:aux:hdrgainmap"][0]
aux_image = heif_file.get_aux_image(aux_id)
```

The reason I avoided loading AUX images during the initialization of `HeifImage` is because we currently don't support all possible aux image formats. So if we simply ignore such aux images, a user might think that the aux image is missing from their file (or at least take some time to realize that it's a feature that `pillow-heif` is missing). But the way I've implemented it now communicates the library's limitations clearly. Do you have an alternate suggestion?

bigcat88 commented:
> I'll add tests soon, and I have a sample heic image that I took on my iphone (having hdrgainmap) and cropped so the size is ~230KB. Shall I simply add it to the repo at `tests/images/heif_other`? (Does this repo use Git LFS?)

No, repo does not use Git LFS. And yes, just add the file to the `tests/images/heif_other`

> The reason I avoided loading AUX images during the initialization of `HeifImage` is because we currently don't support all possible aux image formats. So if we simply ignore such aux images, a user might think that the aux image is missing from their file (or at least take some time to realize that it's a feature that `pillow-heif` is missing). But the way I've implemented it now communicates the library's limitations clearly. Do you have an alternate suggestion?

what you said makes sense, since it's a feature in the beta version, it doesn't have to be implemented for the pillow_plugin.

johncf commented:
One of the images in the repo already had hdrgainmap aux image. So I just used it in the new test.

bigcat88 commented:
As far as I understand, the PR is ready?

If it is not too much trouble, could you also modify this test:

https://github.com/bigcat88/pillow_heif/blob/ba648514fde28dbf4abde8b941caf1894c1a2f16/tests/options_test.py#L27

In theory, three lines should be added there, just like with DEPTH_IMAGES

johncf commented:
> As far as I understand, the PR is ready?

Yes

> three lines should be added there, just like with DEPTH_IMAGES

Done!

bigcat88 reviewed (approved):
Thanks a lot for the great PR, the release will most likely be at the end of the week.
